### PR TITLE
[WIP][RFC][Behat][Promotion] Basic scenarios for promotions

### PR DESCRIPTION
**Pull request by Zales0123**
Hello everybody!

Taxes and shipping related scenarios will be, hopefully, fully merged soon, so it's time to introduce some new stuff :smile:
Here are two simple features about promotions, that can be a base for further, more hardcore scenarios. Lots of steps are already implemented, the only left are those related with promotions logic. For now, there are scenarios for "Fixed discount" and "Percentage discount" promotions, but obviously it's needed to cover as many cases as possible...
And this is main thing to discuss. In there nearest future, we'll try to implement big promotions system improvement (inspired by @peteward https://github.com/Sylius/Sylius/pull/3737). There are a lot of possible promotion combinations and will be even more after this work. We should deeply consider, how many cases do we want to cover with behat.

However, for now there are features that are possible to achieve with current promotions system. I will be glad to hear your thoughts, is *force* strong with these scenarios and is it the way we should go with them ;)

### CODE REVIEW

File: features/shop/apply_correct_simple_promotion_on_order.feature [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    As a Customer
+    I want to have promotions applied
+
+    Background:
+        Given the store is operating on a single channel
+        And default currency is "EUR"
+        And store has a product "PHP T-Shirt" priced at "€100.00"
+        And store has a product "Symfony Mug" priced at "€5.00"
+        And there is promotion "Special discount"
+        And it has "€10.00" "fixed discount" for cart with "item total" above "€80.00"
```
Comment (pamil) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Can we use discount types without wrapping them with quotation marks? `(.+? discount) for cart`

Comment (Zales0123) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: We could think about it, but then we should standardization of discount names :D For example, now we have "Fixed **discount**", "Percentage **discount**" and "Shipping **discount**" actions, but also "Add product" ;)

Comment (pamil) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Anyway, if we don't want to get into trap like with country names transforming to country codes in old contexts, we may even want to create anemic, simple steps for every discount type.

File: features/shop/apply_correct_simple_promotion_on_order.feature [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        And there is promotion "Special discount"
```
Comment (michalmarcinkowski) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Why duplicated step?

File: features/shop/apply_correct_simple_promotion_on_order.feature (hunk 1 above)
Comment (Zales0123) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Sure, we can think about it :+1: Trap is never a good option.
tumblr_inline_nzfevykxev1rg0x34_500

File: features/shop/apply_correct_simple_promotion_on_order.feature (hunk 2 above)
Comment (Zales0123) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Well, not duplicated but badly named, it is a different promotion ;)

File: features/shop/apply_correct_simple_promotion_on_order.feature (hunk 1 above)
Comment (pjedrzejewski) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: This leaks implementation detail and does not sounds natural when you read it. I think we should have steps for different promotions, it is not that big overhead:

```
And store is running promotion "Christmas Sale"
And it gives €10 discount for customers with carts above €80
```

Comment (pjedrzejewski) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Or:

```
And store is running promotion "Christmas Sale"
And it gives 10% discount for customers with cart total above €100
```

**Comment by michalmarcinkowski** [comment]
I would go for something similiar to:
```
And the store has promotion "Get €10 off for order with total greater that €80"
And it has fixed discount of €10.00 for a cart with total greater than €80.00
```

Later we also support discount and criteria split e.g.
```
And the store has promotion "Get €10 off and free shipping for order with total greater that €80"
And it has fixed discount of €10.00
And it has a free shipping discount
And it it applied to a cart with items total greater than €80.00
```

File: features/shop/apply_correct_simple_promotion_on_order.feature (hunk 1 above)
Comment (pjedrzejewski) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: Future scenarios:

```
And store is running promotion "Super-duper Lottery"
And it gives 10% discount for all "Dresses" with a valid coupon
And this promotion has following coupons:
   | code  |
   | X653 |
   | T534 |
And coupon "X653" has been used already
```

And so on, we should stay close to this style IMO.

Comment (pamil) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: @pjedrzejewski :+1:

**Comment by pjedrzejewski** [comment]
@michalmarcinkowski In your example the promotion name contains a lot of information which is irrelevant to how it is configured:

1. I noticed " only on the second look.
2. Easy to get desync between promo name and details
3. Generally confusing to have it duplicated

I get bored after reading 20 steps in the background. I think we should prefer short, promotion specific step definitions over implementing builder pattern in the background. :D

File: features/shop/apply_correct_simple_promotion_on_order.feature (hunk 1 above)
Comment (Zales0123) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: :+1:

Comment (michalmarcinkowski) [inline features/shop/apply_correct_simple_promotion_on_order.feature]: :+1:

**Comment by Zales0123** [comment]
I think that for simple promotion, step like ``And it gives €10 discount for customers with carts above €80`` is really good - we have every information in one line, and it could not be clearer. For more complicated promotions, we could implement a hardly-considered table, as it probably will be more clearer than millions of steps one below another... we should definitely think about ``Background`` size, as it have potential to be a real poem in more specified features ;)

**Comment by pamil** [comment]
Btw. can we extract "simple" promotions to two features? The first one for fixed discount, the second for percentage discount.

**Comment by pjedrzejewski** [comment]
:+1: for splitting them as much as possible.

```
fixed_discount_promotion.feature
percentage_discount_promotion.feature
```

Or perhaps more specific:

```
fixed_discount_promotion_for_carts_with_total_over_X.feature
etc.
```

Not sure about the X thingy, but I can easily get what it means. :D

**Comment by Zales0123** [comment]
Btw, are we ready to use some folders to group features inside ``features/shop/`` etc? :smile: We already have 9 ``.feature`` files inside, 2 two go, only this PR would contain 3 or more... It starts to be quite messy.

**Comment by Zales0123** [comment]
However :+1: for splitting. It is a good question how we should do that, by rule? By action? By rule/action? Features number/Scenarios number balance would be pretty hard to be kept :smile:

**Comment by pjedrzejewski** [comment]
I think it is impossible to come up with a convention that will fit all cases, because you may have combinations of rules, actions, coupons, channels, etc. Let's try to come up with meaningful name for every feature. So:

If we want to describe promotions which have coupons with % discount, we go for:

```
percentage_discount_promotion_applied_with_a_coupon.feature
```

**Comment by Zales0123** [comment]
Ok, so for now I see such features:
- ``apply_fixed_discount_promotion_on_order``
- ``apply_percentage_discount_promotion_on_order``
- ``apply_fixed_discount_promotion_with_taxes_on_order``
- ``apply_percentage_discount_promotion_with_taxes_on_order``

Or shall ``with_taxes`` features be in the *basics* ones? And what with applying multiple promotion? We can handle multiple promotions with the same discount type inside features mentioned above, but there also should be a case with multiple different promotions... ``apply_multiple_different_promotions_on_order``? But there it should be also ``with_taxes`` feature? :smile: In the worst case, there are 6 features for these very simple cases ;/

**Comment by michalmarcinkowski** [comment]
IMO we should divide features to:
 * Verifying promotion rules (I would use a fixed discount action in examples these features):
  * `apply_promotion_based_on_order_total` - scenarios like greater than, equal to, lesser than etc.
  * `apply_promotion_based_on_items_quantity` - scenarios like greater than, equal to, lesser than etc.
  * `apply_promotion_based_on_items_quantity_by_category` - (same as above)
  * `apply_promotion_based_on_order_number` - (same as above)
  * the rest of features for applying single rules
 * Applying promotion actions without any rules
  * `apply_fixed_discount_on_order` - scenarios for edge cases like discount is greater than order total
  * `apply_fixed_discount_on_order_with_shipping_and_taxes` - I would merge verifying shipping and taxes in one feature to reduce the amount of scenarios. We should verify the discount is not applied to shipping and the taxes are correct after the discount is applied

Let's start with one feature, can be `apply_fixed_discount_on_order` and see how this works out. I would avoid writing many features at once, because it will be easier to iterate on them one by one.

**Comment by pjedrzejewski** [comment]
@michalmarcinkowski Makes sense to me. :+1:

**Comment by michalmarcinkowski** [comment]
@Zales0123 I believe this can be closed.

**Comment by Zales0123** [comment]
Yes, I think that everything what should be discussed, have been discussed here ;)

**Pull request closed without merging**